File: sample/src/androidTest/java/com/schibsted/spain/barista/sample/PermissionGranterTest.java
package com.schibsted.spain.barista.sample;

import android.Manifest;
import android.content.Context;
import android.content.pm.PackageManager;
import android.os.Build;
import android.support.annotation.RequiresApi;
import android.support.test.InstrumentationRegistry;
import com.schibsted.spain.barista.BaristaRule;
import com.schibsted.spain.barista.internal.failurehandler.BaristaException;
import com.schibsted.spain.barista.permission.PermissionGranter;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;

import static com.schibsted.spain.barista.BaristaClickActions.click;
import static org.junit.Assume.assumeFalse;
import static org.junit.Assume.assumeTrue;

public class PermissionGranterTest {

  @Rule
  public ExpectedException thrown = ExpectedException.none();

  @Rule
  public BaristaRule<RuntimePermissionActivity> activityRule = BaristaRule.create(RuntimePermissionActivity.class);

  @Before
  public void setUp() throws Exception {
    assumeTrue("This test needs to run on a device with Android 23 or above",
        Build.VERSION.SDK_INT >= Build.VERSION_CODES.M);
    assumeFalse("This test expects you to not have the permission granted. Remember to clear data.",
        hasNeededPermission(InstrumentationRegistry.getTargetContext(), RuntimePermissionActivity.SOME_PERMISSION));
  }

  @Test(expected = BaristaException.class)
  public void fails_when_using_permission() throws Exception {
    activityRule.launchActivity();

    click(R.id.use_permission_button);
  }

  @Test
  @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN_MR2)
  public void works_after_granting_permission() throws Exception {
    activityRule.launchActivity();

    click(R.id.request_permission_button);

    PermissionGranter.allowPermissionsIfNeeded(Manifest.permission.READ_CONTACTS);

    click(R.id.use_permission_button);
  }

  private static boolean hasNeededPermission(Context context, String permissionNeeded) {
    int permissionStatus = context.checkPermission(permissionNeeded, android.os.Process.myPid(), android.os.Process.myUid());
    return permissionStatus == PackageManager.PERMISSION_GRANTED;
  }
}

File: sample/src/androidTest/java/com/schibsted/spain/barista/sample/PermissionGranterTest.kt
package com.schibsted.spain.barista.sample

import android.Manifest
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.test.core.app.ActivityScenario
import com.schibsted.spain.barista.interaction.PermissionGranter
import org.junit.Assert.assertEquals
import org.junit.Assume.assumeTrue
import org.junit.Before
import org.junit.Test

class PermissionGranterTest {

    @Before
    fun setUp() {
        assumeTrue("This test needs to run on a device with Android 23 or above", Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)
    }

    @Test
    fun grants_simple_permission() {
        launchActivity {
            verifyPermissionNotGranted(CONTACTS_PERMISSION)
            requestPermission(CONTACTS_PERMISSION)

            PermissionGranter.allowPermissionsIfNeeded(CONTACTS_PERMISSION)

            verifyPermissionGranted(CONTACTS_PERMISSION)
        }
    }

    /**
     * Since Android Q the Location permission dialog is different to the others
     */
    @Test
    fun grants_location_permission() {
        launchActivity {
            verifyPermissionNotGranted(LOCATION_PERMISSION)
            requestPermission(LOCATION_PERMISSION)

            PermissionGranter.allowPermissionsIfNeeded(LOCATION_PERMISSION)

            verifyPermissionGranted(LOCATION_PERMISSION)
        }
    }

    @Test
    fun ignores_already_granted_permission() {
        launchActivity {
            verifyPermissionNotGranted(CAMERA_PERMISSION)
            requestPermission(CAMERA_PERMISSION)
            PermissionGranter.allowPermissionOneTime(CAMERA_PERMISSION)
            verifyPermissionGranted(CAMERA_PERMISSION)

            PermissionGranter.allowPermissionsIfNeeded(CAMERA_PERMISSION)
        }
    }


}

// We can't reuse permission from one test to another, because they stay granted after each test
private const val CONTACTS_PERMISSION = Manifest.permission.READ_CONTACTS
private const val CAMERA_PERMISSION = Manifest.permission.CAMERA
private const val LOCATION_PERMISSION = Manifest.permission.ACCESS_FINE_LOCATION

private fun ActivityScenario<*>.requestPermission(permission: String) {
    onActivity { activity ->
        ActivityCompat.requestPermissions(activity, arrayOf(permission), 1)
    }
}

private fun ActivityScenario<*>.verifyPermissionNotGranted(permission: String) {
    onActivity { activity ->
        val permissionValue = ContextCompat.checkSelfPermission(activity, permission)
        assertEquals("Permission $permission expected to be denied but was granted;", PackageManager.PERMISSION_DENIED, permissionValue)
    }
}

private fun ActivityScenario<*>.verifyPermissionGranted(permission: String) {
    onActivity { activity ->
        val permissionValue = ContextCompat.checkSelfPermission(activity, permission)
        assertEquals("Expected permission $permission was not granted;", PackageManager.PERMISSION_GRANTED, permissionValue)
    }
}

private fun launchActivity(block: ActivityScenario<*>.() -> Unit) {
    ActivityScenario.launch<EmptyActivity>(EmptyActivity::class.java).use { scenario ->
        scenario.block()
    }
}
